cli/main.rs

```diff
@@ -49,49 +64,252 @@ fn main() {
+                    let mut expr = String::from("");
+
+                    if line.ends_with(".") {
+                        // expr = complete_on.split()
+                    } else {
+                        let bits: Vec<&str> = complete_on.split(".").collect();
+                        // filter = bits.pop().unwrap();
+                        expr = bits.join(".");
+                    }
+
+                    if expr.is_empty() {
+                    } else {
+                        // based on the new node repl, we should probably parse the source into an ESTree
+                        // and then try to get the corresponding propertyNames?
+                    }
                 }
+
+                None => {}
             }
-            Err(ReadlineError::Interrupted) => {
-                println!("Thanks for stopping by!");
-                break;
-            }
-            Err(ReadlineError::Eof) => {
-                println!("Eof'd");
-                break;
-            }
-            Err(err) => {
-                eprintln!("Error: {:?}", err);
-                break;
+            completion_groups
+        }
+    }
+
+    pub fn readline(&mut self) {
+        use rustyline::error::ReadlineError;
+
+        // This looks kinda weird, but Rustyline Editor requires a struct to be passed as Helper.
+        // Since Repl has already impl'd all the traits necessary for Helper we use that instead.
+
+        let helper = RustylineHelper {
+            completions: vec![],
+        };
+
+        let mut rl = Editor::new();
+        rl.set_helper(Some(helper));
+
+        println!("Welcome to jstime v{}!", env!("CARGO_PKG_VERSION"));
+
+        loop {
+            let readline = rl.readline(">> ");
+            match readline {
+                Ok(line) => {
+                    rl.add_history_entry(line.as_str());
+                    match line.as_str() {
+                        ".exit" => {
+                            println!("Thanks for stopping by!");
+                            break;
+                        }
+                        _ => match self.jstime.run_script(&line, ".repl") {
+                            Ok(v) => println!("{}", v),
+                            Err(e) => eprintln!("Uncaught: {}", e),
+                        },
+                    }
+                }
+                Err(ReadlineError::Interrupted) => {
+                    println!("Thanks for stopping by!");
+                    break;
+                }
+                Err(ReadlineError::Eof) => {
+                    println!("Eof'd");
+                    break;
+                }
+                Err(err) => {
+                    eprintln!("Error: {:?}", err);
+                    break;
+                }
             }
         }
+
+        rl.save_history(&self.history_file).unwrap();
     }
+}
+
+#[test]
 
-    if let Some(history_path) = history_path {
-        let _ = rl.save_history(&history_path);
+fn test_global_completions() {
```

Comment:
would be better to test the actual repl interface, like a file in the test dir which spawns jstime cli, sends some characters to stdin, and validates stdout.